// ytmusic/dist/lib/endpoints/HomePage.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spread = (this && this.__spread) || function () {
    for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getFullHomePage = exports.getHomePage = void 0;
var utils = __importStar(require("../utils"));
var continuations_1 = require("../continuations");
var parseTwoRowItemRenderer = utils.parser(function (e) {
    var item = {
        thumbnail: [],
        title: e.title.runs[0],
        subtitle: [],
        navigationEndpoint: e.navigationEndpoint
    };
    e.thumbnailRenderer.musicThumbnailRenderer.thumbnail.thumbnails.forEach(function (el) { return item.thumbnail.push(el); });
    e.subtitle.runs.forEach(function (el) {
        var sub = {
            text: el.text
        };
        if (el.navigationEndpoint) {
            sub.navigationEndpoint = el.navigationEndpoint;
        }
        item.subtitle.push(sub);
    });
    return item;
});
var parseFlexColumnRenderer = utils.parser(function (item) {
    return {
        thumbnail: [],
        title: item.text.runs[0].text,
        subtitle: [],
        navigationEndpoint: item.text.runs[0].navigationEndpoint
    };
});
function parseResponsiveListItemRenderer(e) {
    if (e.flexColumns) {
        return utils.filterMap(e.flexColumns, function (item) {
            if (item.musicResponsiveListItemFlexColumnRenderer) {
                return parseFlexColumnRenderer(item.musicResponsiveListItemFlexColumnRenderer);
            }
            throw new Error("Unexpected flexColumn content: " + JSON.stringify(item));
        });
    }
    throw new Error("Unexpected responsive contents: " + JSON.stringify(e));
}
function parseCarouselItems(e) {
    if (e.musicTwoRowItemRenderer) {
        return [parseTwoRowItemRenderer(e.musicTwoRowItemRenderer)];
    }
    if (e.musicResponsiveListItemRenderer) {
        return parseResponsiveListItemRenderer(e.musicResponsiveListItemRenderer);
    }
    throw new Error("Unexpected carousel contents: " + JSON.stringify(e));
}
var parseCarouselContents = function (contents) {
    return utils.filterMap(contents, function (carousel) {
        if (carousel.musicTastebuilderShelfRenderer)
            return;
        var ctx = carousel.musicCarouselShelfRenderer
            ? carousel.musicCarouselShelfRenderer
            : carousel.musicImmersiveCarouselShelfRenderer;
        var content = utils.filterFlatMap(ctx.contents, parseCarouselItems);
        return {
            title: ctx.header.musicCarouselShelfBasicHeaderRenderer.title.runs[0].text,
            content: content,
            strapline: ctx.header.musicCarouselShelfBasicHeaderRenderer.strapline
                ? ctx.header.musicCarouselShelfBasicHeaderRenderer.strapline.runs
                : undefined
        };
    });
};
/**
 * Returns First Part of HomePage
 *
 * @usage
 *
 * ```js
 *  const api = new YTMUSIC(cookie)
 *  const data = await api.getHomePage()
 * ```
 *
 * You can call `data.continue()` to get next part
 * @returns {@link HomePage}
 *
 */
exports.getHomePage = function (cookie, args) { return __awaiter(void 0, void 0, void 0, function () {
    var response, data, sectionListRenderer, content, home;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0: return [4 /*yield*/, utils.sendRequest(cookie, {
                    id: 'FEmusic_home',
                    endpoint: 'browse',
                    userID: args.userID,
                    authUser: args.authUser
                })];
            case 1:
                response = _a.sent();
                data = response.contents.singleColumnBrowseResultsRenderer.tabs[0].tabRenderer;
                sectionListRenderer = data.content.sectionListRenderer;
                content = parseCarouselContents(sectionListRenderer.contents);
                home = {
                    title: data.title,
                    browseId: data.endpoint.browseEndpoint.browseId,
                    content: content
                };
                home.continue = continuations_1.createContinuation(cookie, args, parseCarouselContents, home, sectionListRenderer);
                return [2 /*return*/, home];
        }
    });
}); };
/**
 * Returns Full HomePage
 *
 * @usage
 * ```js
 *  const api = new YTMUSIC(cookie)
 *  const data = await api.getFullHomePage()
 * ```
 * @returns {@link HomePage}
 *
 */
exports.getFullHomePage = function (cookie, args) { return __awaiter(void 0, void 0, void 0, function () {
    var home, t;
    var _a, _b;
    return __generator(this, function (_c) {
        switch (_c.label) {
            case 0: return [4 /*yield*/, exports.getHomePage(cookie, args)];
            case 1:
                home = _c.sent();
                _c.label = 2;
            case 2:
                if (!true) return [3 /*break*/, 4];
                return [4 /*yield*/, ((_a = home.continue) === null || _a === void 0 ? void 0 : _a.call(home))];
            case 3:
                t = _c.sent();
                if (!t || !t.content)
                    return [3 /*break*/, 4];
                (_b = home.content) === null || _b === void 0 ? void 0 : _b.push.apply(_b, __spread(t.content));
                if (!t.continue)
                    return [3 /*break*/, 4];
                home.continue = t.continue;
                return [3 /*break*/, 2];
            case 4: return [2 /*return*/, home];
        }
    });
}); };
